### PR TITLE
Added functionality for downloading spectrum image

### PR DESCRIPTION
**Pull request by waridrox**
The current `Download` button at http://spectralworkbench.org doesn't work. A blank page is open instead. The added function could be utilised wherever possible for example in the new interface to download the spectrum images with the timestamp. Here's a demo for the same: 

https://user-images.githubusercontent.com/58583793/128475067-207d9b0b-aa51-4c88-8448-e670d6284ade.mov

* [ ] tests pass -- see README.md for how to run them
* [ ] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [ ] pull request are descriptively named
* [ ] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

File: examples/capture/capture.js [hunk 1]
```diff
@@ -161,6 +161,27 @@ $W = {
     this_.getRecentCalibrations("#calibration_id");
   },
 
+  downloadSpectrum: function() {
```
Comment (jywarren) [inline examples/capture/capture.js:164]: This looks awesome. Is this something we ought to write a test for? It seems like it could be a pretty easy test?

Comment (waridrox) [reply examples/capture/capture.js:164]: Sure, in which file would it make more sense tho 😅

Comment (jywarren) [reply examples/capture/capture.js:164]: So the test assertion would be to confirm what `a.href` equals.

Comment (jywarren) [reply examples/capture/capture.js:164]: https://github.com/waridrox/spectral-workbench.js/blob/downloadSpectrum/cypress/integration/spectral-workbench-demo.spec.js

Comment (jywarren) [reply examples/capture/capture.js:164]: Could we try using Cypress to press the download button and then assert if `a.href` is not empty? If that doesn't work, then we can try inserting an image to the <video> element to see if that works?

**Comment by waridrox** [comment]
So I think the `a.href` attribute is actually generated during the click of the download button.

**Comment by jywarren** [comment]
> So I think the `a.href` attribute is actually generated during the click of the download button.

Yes that should be fine, we can simulate a click and then check the href attribute to see if it correctly generated?

**Comment by jywarren** [comment]
Another way you could insert the image is to just overwrite $W.data, I think, before clicking download, but that might test a narrower set of steps. Still useful!

**Comment by waridrox** [comment]
So the anchor tag is generated and destroyed really fast according to the implementation. Even after checking for its presence, the tests prove to be really slow to detect it. 
```javascript 
let a = document.createElement('a');
  a.href = base64_imgdata;
  a.download = ('spectrum_img.png');
  a.click();
  }
```

Same's the case for `$W.data`. If we try to get the current data after clicking the Download button, then that wouldn't work since by that time, the `base64` of the spectrum would be changed to a new one 😕

**Comment by jywarren** [comment]
wait, but why would it be changed to a new one - can't we just pass in a
static image? Then if it returns anything at all, that's not blank, we
should be OK -- or maybe I'm missing something? Thanks!

**Comment by waridrox** [comment]
The data keeps changing due to the default nature of cypress which provides `--use-fake-device-for-media-stream` flag that enables a sample video feed running in the background. 
Screenshot 2021-08-22 at 6 31 48 PM


This means we can't assert the `base64 data` as it is since that will change really fast - https://github.com/cypress-io/cypress/issues/2704

The other thing is that the cypress tests stop since there is a prompt for downloading the `spectrum_img`. 



https://user-images.githubusercontent.com/58583793/130356720-d4d8249e-7555-496e-8a78-6b28fbe6c3aa.mp4

**Comment by jywarren** [comment]
However, we can set a test video with `--use-fake-device-for-media-stream` manually and make it one which does not change -- then we can predict the displayed values. What do you think? Some very simple video with all same frames?

**Comment by jywarren** [comment]
> The other thing is that the cypress tests stop since there is a prompt for downloading the spectrum_img.

I think there ought to be a cypress tool for accepting the prompt... https://testersdock.com/cypress-javascript-alert-confirm-prompt/

**Comment by waridrox** [comment]
> I think there ought to be a cypress tool for accepting the prompt... https://testersdock.com/cypress-javascript-alert-confirm-prompt/

Sure, this prompt however I think is not a JS generated so there are means to get passed this like - https://www.npmjs.com/package/cypress-downloadfile / Cypress v6.xx

**Comment by jywarren** [comment]
Ahhh this prompt -- it's an HTML feature, so there should be a cypress way around it, OR maybe we could just click and then compare the `a.download` attribute...

**Comment by jywarren** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/HTMLAnchorElement/download

https://stackoverflow.com/a/66811222

> Cypress can easily test downloadable files [since version 6.3.0](https://github.com/cypress-io/cypress/issues/5027#issuecomment-763325232)

Also possibly -- https://github.com/cypress-io/cypress/issues/949

**Comment by jywarren** [comment]
Hi @waridrox would you like me to give this one a try? Happy to if you'd like!

**Comment by waridrox** [comment]
Hi @jywarren!, I’m currently busy with my exams right now for the upcoming week or two 😅. If you would like to experiment, please by all means go ahead 🙌

**Comment by jywarren** [comment]
OK, will give it a try if I have a moment - good luck in your exams!!!

File: examples/new-capture/index.html [hunk 2]
```diff
@@ -246,9 +246,14 @@ <h1>
           </div>
         </div>
 
+        <input name="dataurl" type="hidden" id="dataurl" />
+
         <button class="demo-button next" id="capture-page-next">Save Capture</button>
         <p style="margin-top: 0.5em">Once you save the capture, you cannot go back here.</p>
         <img style="display:none;background:#333;" id="spectrum-preview" />
+
+        <button class="demo-button next" id="download-spectrum" onClick="$W.downloadSpectrum();">Download</button>
```
Comment (jywarren) [inline examples/new-capture/index.html:255]: here's the button

**Comment by jywarren** [comment]
https://github.com/cypress-io/cypress-example-recipes/tree/master/examples/testing-dom__download

**Comment by jywarren** [comment]
https://github.com/cypress-io/cypress-example-recipes/blob/2e5fb3da420f544637ff77d81ff7261f5d2164e8/examples/testing-dom__download/cypress/integration/local-download-spec.js#L19-L28

**Comment by jywarren** [comment]
Let's start by testing that it's /some/ minimum length. Then we can also try presetting the video output using https://github.com/publiclab/spectral-workbench/issues/710 for a more specific test.

**Comment by jywarren** [comment]
```

  Actions
    ✓ It reach the default landing page (1421ms)
    ✓ The header navigation stepper is present and it is in the correct order (422ms)
    ✓ The default Home page step should be highlighted but not the rest (327ms)
    1) can be clicked through to begin capturing


  3 passing (4s)
  1 failing

  1) Actions
       can be clicked through to begin capturing:
     ReferenceError: path is not defined
      at Context.eval (http://127.0.0.1:8080/__cypress/tests?p=cypress/integration/spectral-workbench-demo.spec.js:128:22)
```

File: cypress/integration/spectral-workbench-demo.spec.js [hunk 3]
```diff
@@ -28,5 +28,23 @@ context('Actions', () => {
...
+    cy.log('**read downloaded file**')
+
+    // file path is relative to the working folder
+    const filename = path.join(downloadsFolder, 'spectrum_img.png')
+
+    // browser might take a while to download the file,
+    // so use "cy.readFile" to retry until the file exists
+    // and has length - and we assume that it has finished downloading then
+    cy.readFile(filename, { timeout: 15000 })
+    .should('have.length.gt', 5000000000)
```
Comment (jywarren) [inline cypress/integration/spectral-workbench-demo.spec.js:47]: This failed. Great!

Comment (jywarren) [inline cypress/integration/spectral-workbench-demo.spec.js:47]:
```suggestion
    .should('have.length.gt', 50)
```

**Comment by jywarren** [comment]
Nice!!! Woohoo!!

**Pull request merged**